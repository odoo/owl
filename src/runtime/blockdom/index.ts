export { config } from "./config";

export { toggler } from "./toggler";
export { createBlock } from "./block_compiler";
export { list } from "./list";
export { multi } from "./multi";
export { text, comment } from "./text";
export { html } from "./html";
export { createCatcher } from "./event_catcher";

export interface VNode<T = any> {
  mount(parent: HTMLElement, afterNode: Node | null): void;
  moveBeforeDOMNode(node: Node | null, parent?: HTMLElement): void;
  moveBeforeVNode(other: T | null, afterNode: Node | null): void;
  patch(other: T, withBeforeRemove: boolean): void;
  beforeRemove(): void;
  remove(): void;
  firstNode(): Node | undefined;

  el?: undefined | HTMLElement | Text;
  parentEl?: undefined | HTMLElement;
  isOnlyChild?: boolean | undefined;
  key?: any;
}

export type BDom = VNode<any>;

export function mount(vnode: VNode, fixture: HTMLElement, afterNode: Node | null = null) {
  vnode.mount(fixture, afterNode);
}

export function patch(vnode1: VNode, vnode2: VNode, withBeforeRemove: boolean = false) {
  vnode1.patch(vnode2, withBeforeRemove);
}

export function remove(vnode: VNode, withBeforeRemove: boolean = false) {
  if (withBeforeRemove) {
    vnode.beforeRemove();
  }
  vnode.remove();
}

export function withKey(vnode: VNode, key: any) {
  vnode.key = key;
  return vnode;
}
