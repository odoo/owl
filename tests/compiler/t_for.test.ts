import {
  renderToBdom,
  renderToString,
  snapshotEverything,
  TestContext,
  makeTestFixture,
} from "../helpers";
import { mount, patch } from "../../src/runtime/blockdom";

snapshotEverything();

// -----------------------------------------------------------------------------
// t-for
// -----------------------------------------------------------------------------

describe("t-for", () => {
  test("simple iteration", () => {
    const template = `<t t-for="item" t-of="[3, 2, 1]" t-key="item"><t t-esc="item"/></t>`;
    expect(renderToString(template)).toBe("321");
  });

  test("simple iteration with two nodes inside", () => {
    const template = `
      <t t-for="item" t-of="[3, 2, 1]" t-key="item">
        <span>a<t t-esc="item"/></span>
        <span>b<t t-esc="item"/></span>
      </t>`;
    const expected =
      "<span>a3</span><span>b3</span><span>a2</span><span>b2</span><span>a1</span><span>b1</span>";
    expect(renderToString(template)).toBe(expected);
  });

  test("destructuring array items", () => {
    const template = `<t t-for="[key, value]" t-of="Object.entries({ a: 1, b: 2 })" t-key="key">(<t t-esc="key"/>: <t t-esc="value"/>)</t>`;
    expect(renderToString(template)).toBe("(a: 1)(b: 2)");
  });

  test("destructuring array items: rest", () => {
    const template = `<t t-for="[head, ...tail]" t-of="[[1, 2, 3], [4, 5, 6]]" t-key="head">(<t t-esc="head"/>;<t t-esc="tail"/>)</t>`;
    expect(renderToString(template)).toBe("(1;2,3)(4;5,6)");
  });

  test("destructuring object items", () => {
    const template = `<t t-for="{ k, v }" t-of="[{ k: 'a', v: 1 }, { k: 'b', v: 2 }]" t-key="k">(<t t-esc="k"/>: <t t-esc="v"/>)</t>`;
    expect(renderToString(template)).toBe("(a: 1)(b: 2)");
  });

  test("destructuring object items: rest", () => {
    const template = `<t t-for="{ k, v }" t-of="[{ k: 'a', v: 1 }, { k: 'b', v: 2 }]" t-key="k">(<t t-esc="k"/>: <t t-esc="v"/>)</t>`;
    expect(renderToString(template)).toBe("(a: 1)(b: 2)");
  });

  test("nested destructuring", () => {
    const template = `<t t-for="[key, {left, right}]" t-of="Object.entries(obj)" t-key="key">(<t t-esc="key"/>: [<t t-esc="left"/>, <t t-esc="right"/>])</t>`;
    expect(
      renderToString(template, {
        obj: {
          a: { left: 1, right: 2 },
          b: { left: 3, right: 4 },
        },
      })
    ).toBe("(a: [1, 2])(b: [3, 4])");
  });

  test("t-key on t-for", async () => {
    const template = `
        <div>
          <t t-for="thing" t-of="things" t-key="thing">
            <span/>
          </t>
        </div>`;

    const fixture = makeTestFixture();

    const vnode1 = renderToBdom(template, { things: [1, 2] });
    mount(vnode1, fixture);
    let elm = fixture;
    expect(elm.innerHTML).toBe("<div><span></span><span></span></div>");
    const first = elm.querySelectorAll("span")[0];
    const second = elm.querySelectorAll("span")[1];

    const vnode2 = renderToBdom(template, { things: [2, 1] });
    patch(vnode1, vnode2);

    expect(elm.innerHTML).toBe("<div><span></span><span></span></div>");
    expect(first).toBe(elm.querySelectorAll("span")[1]);
    expect(second).toBe(elm.querySelectorAll("span")[0]);
  });

  test("simple iteration (in a node)", () => {
    const template = `
        <div>
          <t t-for="item" t-of="[3, 2, 1]" t-key="item"><t t-esc="item"/></t>
        </div>`;
    expect(renderToString(template)).toBe("<div>321</div>");
  });

  test("iterate on items (on a element node)", () => {
    const template = `
        <div>
          <span t-for="item" t-of="[1, 2]" t-key="item"><t t-esc="item"/></span>
        </div>`;
    const expected = `<div><span>1</span><span>2</span></div>`;
    expect(renderToString(template)).toBe(expected);
  });

  test("iterate, Map param", () => {
    const template = `
      <t t-for="[key, value]" t-of="map" t-key="key">
        [<t t-esc="key"/>: <t t-esc="value"/>]
      </t>`;
    const expected = ` [a: 1]  [b: 2]  [c: 3] `;
    const context = {
      map: new Map([
        ["a", 1],
        ["b", 2],
        ["c", 3],
      ]),
    };
    expect(renderToString(template, context)).toBe(expected);
  });

  test("iterate, Set param", () => {
    const template = `
      <t t-for="item" t-of="set" t-key="item">
        <t t-esc="item"/>
      </t>`;
    const expected = `123`;
    const context = { set: new Set([1, 2, 3]) };
    expect(renderToString(template, context)).toBe(expected);
  });

  test("iterate, iterable param", () => {
    const template = `
      <t t-for="item" t-of="map.values()" t-key="item">
        <t t-esc="item"/>
      </t>`;
    const expected = `123`;
    const context = {
      map: new Map([
        ["a", 1],
        ["b", 2],
        ["c", 3],
      ]),
    };
    expect(renderToString(template, context)).toBe(expected);
  });

  test("iterate, generator param", () => {
    const template = `
      <t t-for="item" t-of="gen()" t-key="item">
        <t t-esc="item"/>
      </t>`;
    const expected = `123`;
    const context = {
      *gen() {
        yield 1;
        yield 2;
        yield 3;
      },
    };
    expect(renderToString(template, context)).toBe(expected);
  });

  test("does not pollute the rendering context", () => {
    const template = `
        <div>
          <t t-for="item" t-of="[1]" t-key="item"><t t-esc="item"/></t>
        </div>`;
    const context = { __owl__: {} };
    renderToString(template, context);
    expect(Object.keys(context)).toEqual(["__owl__"]);
  });

  test("t-for in t-for", () => {
    const template = `
        <div>
          <t t-for="number" t-of="numbers" t-key="number">
            <t t-for="letter" t-of="letters" t-key="letter">
              [<t t-esc="number"/><t t-esc="letter"/>]
            </t>
          </t>
        </div>`;

    const context = { numbers: [1, 2, 3], letters: ["a", "b"] };
    const expected = "<div> [1a]  [1b]  [2a]  [2b]  [3a]  [3b] </div>";
    expect(renderToString(template, context)).toBe(expected);
  });

  test("t-for in t-foreach", () => {
    const template = `
        <div>
          <t t-foreach="numbers" t-as="number" t-key="number">
            <t t-for="letter" t-of="letters" t-key="letter">
              [<t t-esc="number"/><t t-esc="letter"/>]
            </t>
          </t>
        </div>`;

    const context = { numbers: [1, 2, 3], letters: ["a", "b"] };
    const expected = "<div> [1a]  [1b]  [2a]  [2b]  [3a]  [3b] </div>";
    expect(renderToString(template, context)).toBe(expected);
  });

  test("t-foreach in t-for", () => {
    const template = `
        <div>
          <t t-for="number" t-of="numbers" t-key="number">
            <t t-foreach="letters" t-as="letter" t-key="letter">
              [<t t-esc="number"/><t t-esc="letter"/>]
            </t>
          </t>
        </div>`;

    const context = { numbers: [1, 2, 3], letters: ["a", "b"] };
    const expected = "<div> [1a]  [1b]  [2a]  [2b]  [3a]  [3b] </div>";
    expect(renderToString(template, context)).toBe(expected);
  });

  test("t-call without body in t-for in t-for", () => {
    const context = new TestContext();
    const sub = `
        <t>
          <t t-set="c" t-value="'x' + '_' + a + '_'+ b" />
          [<t t-esc="a" />]
          [<t t-esc="b" />]
          [<t t-esc="c" />]
        </t>`;

    const main = `
        <div>
          <t t-for="a" t-of="numbers" t-key="a">
            <t t-for="b" t-of="letters" t-key="b">
              <t t-call="sub" />
            </t>
            <span t-esc="c"/>
          </t>
          <span>[<t t-esc="a" />][<t t-esc="b" />][<t t-esc="c" />]</span>
        </div>`;

    context.addTemplate("sub", sub);
    context.addTemplate("main", main);

    const ctx = { numbers: [1, 2, 3], letters: ["a", "b"] };
    const expected =
      "<div> [1] [a] [x_1_a]  [1] [b] [x_1_b] <span></span> [2] [a] [x_2_a]  [2] [b] [x_2_b] <span></span> [3] [a] [x_3_a]  [3] [b] [x_3_b] <span></span><span>[][][]</span></div>";
    expect(context.renderToString("main", ctx)).toBe(expected);
  });

  test("t-call with body in t-for in t-for", () => {
    const context = new TestContext();
    const sub = `
        <t>
          [<t t-esc="a" />]
          [<t t-esc="b" />]
          [<t t-esc="c" />]
        </t>`;

    const main = `
        <div>
          <t t-for="a" t-of="numbers" t-key="a">
            <t t-for="b" t-of="letters" t-key="b">
              <t t-call="sub" >
                <t t-set="c" t-value="'x' + '_' + a + '_'+ b" />
              </t>
            </t>
            <span t-esc="c"/>
          </t>
          <span>[<t t-esc="a" />][<t t-esc="b" />][<t t-esc="c" />]</span>
        </div>`;

    context.addTemplate("sub", sub);
    context.addTemplate("main", main);

    const ctx = { numbers: [1, 2, 3], letters: ["a", "b"] };
    const expected =
      "<div> [1] [a] [x_1_a]  [1] [b] [x_1_b] <span></span> [2] [a] [x_2_a]  [2] [b] [x_2_b] <span></span> [3] [a] [x_3_a]  [3] [b] [x_3_b] <span></span><span>[][][]</span></div>";
    expect(context.renderToString("main", ctx)).toBe(expected);
  });

  test("throws error if invalid loop expression", () => {
    const test = `<div><t t-for="item" t-of="abc" t-key="item"><span t-key="item"/></t></div>`;
    expect(() => renderToString(test)).toThrow("ctx.abc is not iterable");
  });

  test("t-for with t-if inside", () => {
    const template = `
        <div>
          <t t-for="{ id, text }" t-of="elems" t-key="id">
            <span t-if="id lt 3"><t t-esc="text"/></span>
          </t>
        </div>`;
    const ctx = {
      elems: [
        { id: 1, text: "a" },
        { id: 2, text: "b" },
        { id: 3, text: "c" },
      ],
    };
    expect(renderToString(template, ctx)).toBe("<div><span>a</span><span>b</span></div>");
  });

  test("t-for with t-if inside (no external node)", () => {
    const template = `
          <t t-for="{ id, text }" t-of="elems" t-key="id">
            <span t-if="id lt 3"><t t-esc="text"/></span>
          </t>`;
    const ctx = {
      elems: [
        { id: 1, text: "a" },
        { id: 2, text: "b" },
        { id: 3, text: "c" },
      ],
    };
    expect(renderToString(template, ctx)).toBe("<span>a</span><span>b</span>");
  });
});
