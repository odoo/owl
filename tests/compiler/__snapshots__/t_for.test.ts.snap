// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`t-for destructuring array items 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block1 = [];
    let i1 = 0;
    for ([ctx['key'],ctx['value']] of Object.entries({a:1,b:2})) {
      const key1 = ctx['key'];
      const b3 = text(\`(\`);
      const b4 = text(ctx['key']);
      const b5 = text(\`: \`);
      const b6 = text(ctx['value']);
      const b7 = text(\`)\`);
      c_block1[i1] = withKey(multi([b3, b4, b5, b6, b7]), key1);
      i1++;
    }
    return list(c_block1);
  }
}"
`;

exports[`t-for destructuring array items: rest 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block1 = [];
    let i1 = 0;
    for ([ctx['head'],...ctx['tail']] of [[1,2,3],[4,5,6]]) {
      const key1 = ctx['head'];
      const b3 = text(\`(\`);
      const b4 = text(ctx['head']);
      const b5 = text(\`;\`);
      const b6 = text(ctx['tail']);
      const b7 = text(\`)\`);
      c_block1[i1] = withKey(multi([b3, b4, b5, b6, b7]), key1);
      i1++;
    }
    return list(c_block1);
  }
}"
`;

exports[`t-for destructuring object items 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block1 = [];
    let i1 = 0;
    for ({k:ctx['k'],v:ctx['v']} of [{k:'a',v:1},{k:'b',v:2}]) {
      const key1 = ctx['k'];
      const b3 = text(\`(\`);
      const b4 = text(ctx['k']);
      const b5 = text(\`: \`);
      const b6 = text(ctx['v']);
      const b7 = text(\`)\`);
      c_block1[i1] = withKey(multi([b3, b4, b5, b6, b7]), key1);
      i1++;
    }
    return list(c_block1);
  }
}"
`;

exports[`t-for destructuring object items: rest 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block1 = [];
    let i1 = 0;
    for ({k:ctx['k'],v:ctx['v']} of [{k:'a',v:1},{k:'b',v:2}]) {
      const key1 = ctx['k'];
      const b3 = text(\`(\`);
      const b4 = text(ctx['k']);
      const b5 = text(\`: \`);
      const b6 = text(ctx['v']);
      const b7 = text(\`)\`);
      c_block1[i1] = withKey(multi([b3, b4, b5, b6, b7]), key1);
      i1++;
    }
    return list(c_block1);
  }
}"
`;

exports[`t-for does not pollute the rendering context 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block2 = [];
    let i1 = 0;
    for (ctx['item'] of [1]) {
      const key1 = ctx['item'];
      c_block2[i1] = withKey(text(ctx['item']), key1);
      i1++;
    }
    const b2 = list(c_block2);
    return block1([], [b2]);
  }
}"
`;

exports[`t-for iterate on items (on a element node) 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  let block3 = createBlock(\`<span><block-text-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block2 = [];
    let i1 = 0;
    for (ctx['item'] of [1,2]) {
      const key1 = ctx['item'];
      let txt1 = ctx['item'];
      c_block2[i1] = withKey(block3([txt1]), key1);
      i1++;
    }
    const b2 = list(c_block2);
    return block1([], [b2]);
  }
}"
`;

exports[`t-for iterate, Map param 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block1 = [];
    let i1 = 0;
    for ([ctx['key'],ctx['value']] of ctx['map']) {
      const key1 = ctx['key'];
      const b3 = text(\` [\`);
      const b4 = text(ctx['key']);
      const b5 = text(\`: \`);
      const b6 = text(ctx['value']);
      const b7 = text(\`] \`);
      c_block1[i1] = withKey(multi([b3, b4, b5, b6, b7]), key1);
      i1++;
    }
    return list(c_block1);
  }
}"
`;

exports[`t-for iterate, Set param 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block1 = [];
    let i1 = 0;
    for (ctx['item'] of ctx['set']) {
      const key1 = ctx['item'];
      c_block1[i1] = withKey(text(ctx['item']), key1);
      i1++;
    }
    return list(c_block1);
  }
}"
`;

exports[`t-for iterate, generator param 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block1 = [];
    let i1 = 0;
    for (ctx['item'] of ctx['gen']()) {
      const key1 = ctx['item'];
      c_block1[i1] = withKey(text(ctx['item']), key1);
      i1++;
    }
    return list(c_block1);
  }
}"
`;

exports[`t-for iterate, iterable param 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block1 = [];
    let i1 = 0;
    for (ctx['item'] of ctx['map'].values()) {
      const key1 = ctx['item'];
      c_block1[i1] = withKey(text(ctx['item']), key1);
      i1++;
    }
    return list(c_block1);
  }
}"
`;

exports[`t-for nested destructuring 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block1 = [];
    let i1 = 0;
    for ([ctx['key'],{left:ctx['left'],right:ctx['right']}] of Object.entries(ctx['obj'])) {
      const key1 = ctx['key'];
      const b3 = text(\`(\`);
      const b4 = text(ctx['key']);
      const b5 = text(\`: [\`);
      const b6 = text(ctx['left']);
      const b7 = text(\`, \`);
      const b8 = text(ctx['right']);
      const b9 = text(\`])\`);
      c_block1[i1] = withKey(multi([b3, b4, b5, b6, b7, b8, b9]), key1);
      i1++;
    }
    return list(c_block1);
  }
}"
`;

exports[`t-for simple iteration (in a node) 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block2 = [];
    let i1 = 0;
    for (ctx['item'] of [3,2,1]) {
      const key1 = ctx['item'];
      c_block2[i1] = withKey(text(ctx['item']), key1);
      i1++;
    }
    const b2 = list(c_block2);
    return block1([], [b2]);
  }
}"
`;

exports[`t-for simple iteration 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block1 = [];
    let i1 = 0;
    for (ctx['item'] of [3,2,1]) {
      const key1 = ctx['item'];
      c_block1[i1] = withKey(text(ctx['item']), key1);
      i1++;
    }
    return list(c_block1);
  }
}"
`;

exports[`t-for simple iteration with two nodes inside 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  let block3 = createBlock(\`<span>a<block-text-0/></span>\`);
  let block4 = createBlock(\`<span>b<block-text-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block1 = [];
    let i1 = 0;
    for (ctx['item'] of [3,2,1]) {
      const key1 = ctx['item'];
      let txt1 = ctx['item'];
      const b3 = block3([txt1]);
      let txt2 = ctx['item'];
      const b4 = block4([txt2]);
      c_block1[i1] = withKey(multi([b3, b4]), key1);
      i1++;
    }
    return list(c_block1);
  }
}"
`;

exports[`t-for t-call with body in t-for in t-for 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { isBoundary, withDefault, setContextValue, withKey } = helpers;
  const callTemplate_1 = app.getTemplate(\`sub\`);
  
  let block1 = createBlock(\`<div><block-child-0/><span>[<block-text-0/>][<block-text-1/>][<block-text-2/>]</span></div>\`);
  let block6 = createBlock(\`<span><block-text-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    ctx[isBoundary] = 1
    ctx = Object.create(ctx);
    const c_block2 = [];
    let i1 = 0;
    for (ctx['a'] of ctx['numbers']) {
      const key1 = ctx['a'];
      ctx = Object.create(ctx);
      const c_block4 = [];
      let i2 = 0;
      for (ctx['b'] of ctx['letters']) {
        const key2 = ctx['b'];
        ctx = Object.create(ctx);
        ctx[isBoundary] = 1;
        setContextValue(ctx, \\"c\\", 'x'+'_'+ctx['a']+'_'+ctx['b']);
        c_block4[i2] = withKey(callTemplate_1.call(this, ctx, node, key + \`__1__\${key1}__\${key2}\`), key2);
        ctx = ctx.__proto__;
        i2++;
      }
      ctx = ctx.__proto__;
      const b4 = list(c_block4);
      let txt1 = ctx['c'];
      const b6 = block6([txt1]);
      c_block2[i1] = withKey(multi([b4, b6]), key1);
      i1++;
    }
    ctx = ctx.__proto__;
    const b2 = list(c_block2);
    let txt2 = ctx['a'];
    let txt3 = ctx['b'];
    let txt4 = ctx['c'];
    return block1([txt2, txt3, txt4], [b2]);
  }
}"
`;

exports[`t-for t-call with body in t-for in t-for 2`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  return function template(ctx, node, key = \\"\\") {
    const b2 = text(\` [\`);
    const b3 = text(ctx['a']);
    const b4 = text(\`] [\`);
    const b5 = text(ctx['b']);
    const b6 = text(\`] [\`);
    const b7 = text(ctx['c']);
    const b8 = text(\`] \`);
    return multi([b2, b3, b4, b5, b6, b7, b8]);
  }
}"
`;

exports[`t-for t-call without body in t-for in t-for 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  const callTemplate_1 = app.getTemplate(\`sub\`);
  
  let block1 = createBlock(\`<div><block-child-0/><span>[<block-text-0/>][<block-text-1/>][<block-text-2/>]</span></div>\`);
  let block6 = createBlock(\`<span><block-text-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block2 = [];
    let i1 = 0;
    for (ctx['a'] of ctx['numbers']) {
      const key1 = ctx['a'];
      ctx = Object.create(ctx);
      const c_block4 = [];
      let i2 = 0;
      for (ctx['b'] of ctx['letters']) {
        const key2 = ctx['b'];
        c_block4[i2] = withKey(callTemplate_1.call(this, ctx, node, key + \`__1__\${key1}__\${key2}\`), key2);
        i2++;
      }
      ctx = ctx.__proto__;
      const b4 = list(c_block4);
      let txt1 = ctx['c'];
      const b6 = block6([txt1]);
      c_block2[i1] = withKey(multi([b4, b6]), key1);
      i1++;
    }
    ctx = ctx.__proto__;
    const b2 = list(c_block2);
    let txt2 = ctx['a'];
    let txt3 = ctx['b'];
    let txt4 = ctx['c'];
    return block1([txt2, txt3, txt4], [b2]);
  }
}"
`;

exports[`t-for t-call without body in t-for in t-for 2`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { isBoundary, withDefault, setContextValue } = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    ctx[isBoundary] = 1
    setContextValue(ctx, \\"c\\", 'x'+'_'+ctx['a']+'_'+ctx['b']);
    const b2 = text(\` [\`);
    const b3 = text(ctx['a']);
    const b4 = text(\`] [\`);
    const b5 = text(ctx['b']);
    const b6 = text(\`] [\`);
    const b7 = text(ctx['c']);
    const b8 = text(\`] \`);
    return multi([b2, b3, b4, b5, b6, b7, b8]);
  }
}"
`;

exports[`t-for t-for in t-for 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block2 = [];
    let i1 = 0;
    for (ctx['number'] of ctx['numbers']) {
      const key1 = ctx['number'];
      ctx = Object.create(ctx);
      const c_block3 = [];
      let i2 = 0;
      for (ctx['letter'] of ctx['letters']) {
        const key2 = ctx['letter'];
        const b5 = text(\` [\`);
        const b6 = text(ctx['number']);
        const b7 = text(ctx['letter']);
        const b8 = text(\`] \`);
        c_block3[i2] = withKey(multi([b5, b6, b7, b8]), key2);
        i2++;
      }
      ctx = ctx.__proto__;
      c_block2[i1] = withKey(list(c_block3), key1);
      i1++;
    }
    const b2 = list(c_block2);
    return block1([], [b2]);
  }
}"
`;

exports[`t-for t-for in t-foreach 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { prepareList, withKey } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const [k_block2, v_block2, l_block2, c_block2] = prepareList(ctx['numbers']);;
    for (let i1 = 0; i1 < l_block2; i1++) {
      ctx[\`number\`] = v_block2[i1];
      const key1 = ctx['number'];
      ctx = Object.create(ctx);
      const c_block3 = [];
      let i2 = 0;
      for (ctx['letter'] of ctx['letters']) {
        const key2 = ctx['letter'];
        const b5 = text(\` [\`);
        const b6 = text(ctx['number']);
        const b7 = text(ctx['letter']);
        const b8 = text(\`] \`);
        c_block3[i2] = withKey(multi([b5, b6, b7, b8]), key2);
        i2++;
      }
      ctx = ctx.__proto__;
      c_block2[i1] = withKey(list(c_block3), key1);
    }
    const b2 = list(c_block2);
    return block1([], [b2]);
  }
}"
`;

exports[`t-for t-for with t-if inside (no external node) 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  let block3 = createBlock(\`<span><block-text-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block1 = [];
    let i1 = 0;
    for ({id:ctx['id'],text:ctx['text']} of ctx['elems']) {
      const key1 = ctx['id'];
      let b3;
      if (ctx['id']<3) {
        let txt1 = ctx['text'];
        b3 = block3([txt1]);
      }
      c_block1[i1] = withKey(multi([b3]), key1);
      i1++;
    }
    return list(c_block1);
  }
}"
`;

exports[`t-for t-for with t-if inside 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  let block4 = createBlock(\`<span><block-text-0/></span>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block2 = [];
    let i1 = 0;
    for ({id:ctx['id'],text:ctx['text']} of ctx['elems']) {
      const key1 = ctx['id'];
      let b4;
      if (ctx['id']<3) {
        let txt1 = ctx['text'];
        b4 = block4([txt1]);
      }
      c_block2[i1] = withKey(multi([b4]), key1);
      i1++;
    }
    const b2 = list(c_block2);
    return block1([], [b2]);
  }
}"
`;

exports[`t-for t-foreach in t-for 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { prepareList, withKey } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block2 = [];
    let i1 = 0;
    for (ctx['number'] of ctx['numbers']) {
      const key1 = ctx['number'];
      ctx = Object.create(ctx);
      const [k_block3, v_block3, l_block3, c_block3] = prepareList(ctx['letters']);;
      for (let i2 = 0; i2 < l_block3; i2++) {
        ctx[\`letter\`] = v_block3[i2];
        const key2 = ctx['letter'];
        const b5 = text(\` [\`);
        const b6 = text(ctx['number']);
        const b7 = text(ctx['letter']);
        const b8 = text(\`] \`);
        c_block3[i2] = withKey(multi([b5, b6, b7, b8]), key2);
      }
      ctx = ctx.__proto__;
      c_block2[i1] = withKey(list(c_block3), key1);
      i1++;
    }
    const b2 = list(c_block2);
    return block1([], [b2]);
  }
}"
`;

exports[`t-for t-key on t-for 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  let block3 = createBlock(\`<span/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block2 = [];
    let i1 = 0;
    for (ctx['thing'] of ctx['things']) {
      const key1 = ctx['thing'];
      c_block2[i1] = withKey(block3(), key1);
      i1++;
    }
    const b2 = list(c_block2);
    return block1([], [b2]);
  }
}"
`;

exports[`t-for throws error if invalid loop expression 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  let { withKey } = helpers;
  
  let block1 = createBlock(\`<div><block-child-0/></div>\`);
  let block3 = createBlock(\`<span/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    ctx = Object.create(ctx);
    const c_block2 = [];
    let i1 = 0;
    for (ctx['item'] of ctx['abc']) {
      const key1 = ctx['item'];
      const tKey_1 = ctx['item'];
      c_block2[i1] = withKey(block3(), tKey_1 + key1);
      i1++;
    }
    const b2 = list(c_block2);
    return block1([], [b2]);
  }
}"
`;
